components/channel/channel.c

```diff
@@ -0,0 +1,178 @@
+#include "channel.h"
+
+#include <queue/queue.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <uv.h>
+
+struct datadog_php_channel_impl_s {
+    datadog_php_queue queue;
+
+    // enforce single consumer (refcount should be 0 or 1)
+    uint32_t receiver_count;
+
+    /* Multiple producers are fine. When there aren't any items in the queue, the
+     * consumer will use this info to determine whether to try waiting for more
+     * data or to return empty handed. If there is a known producer, it will wait;
+     * otherwise it will return without waiting.
+     */
+    uint32_t sender_count;
+
+    uv_mutex_t mutex;
+    uv_cond_t condvar;
+    void *buffer[];
+};
+
+/**
+ * Destroys the channel. _NOT_ thread safe.
+ */
+static void channel_dtor(datadog_php_channel_impl *channel) {
+    uv_mutex_destroy(&channel->mutex);
+    uv_cond_destroy(&channel->condvar);
+    free(channel);
+}
+
+static bool receiver_recv(struct datadog_php_receiver_s *receiver, void **data, uint64_t timeout_nanos) {
+    if (!receiver || !receiver->channel) {
+        return false;
+    }
+
+    datadog_php_channel_impl *channel = receiver->channel;
+    uv_mutex_lock(&channel->mutex);
+    datadog_php_queue *queue = &channel->queue;
+    bool succeeded = queue->try_pop(queue, data);
+
+    // If there wasn't an item but there are known producers, wait and try again
+    if (!succeeded && channel->sender_count && timeout_nanos) {
+        uint64_t now = uv_hrtime();
+        uint64_t timeout_target = now + timeout_nanos;
+        do {
+            // Handle being signalled, waking up spuriously, and timing out the same.
+            (void)uv_cond_timedwait(&channel->condvar, &channel->mutex, timeout_target - now);
+            succeeded = queue->try_pop(queue, data);
```

Comment:
How is this ever supposed to succeed, given that we hold a lock here? Preventing anyone from modifying the queue (sender_send being properly locked).

Reply:
In fact, as we have the assumption that there will only be one receiver at most, we can first check for availability on the queue _without_ holding a lock. Checking `queue->size` and only acquiring a lock when actually popping from the queue.

Reply:
The docs for `uv_cond_timedwait` are a [bit sparse](http://docs.libuv.org/en/v1.x/threading.html#c.uv_cond_timedwait) but as you can imagine it calls `pthread_cond_timedwait` which [is documented](https://linux.die.net/man/3/pthread_cond_wait):

> The pthread_cond_timedwait() and pthread_cond_wait() functions shall block on a condition variable. They shall be called with mutex locked by the calling thread or undefined behavior results.
>
> These functions atomically release mutex and cause the calling thread to block on the condition variable cond; atomically here means "atomically with respect to access by another thread to the mutex and then the condition variable".

So the lock is released when it waits, and re-acquired when it wakes.